File: agent/frontend/src/api/analysis.ts
import request from './request'
import type { ApiResponse } from '../types/api'

/**
 * 性能瓶颈类型
 */
export interface PerformanceBottleneck {
  /** 目标资源，如 Pod 名称、节点名称等 */
  target: string
  /** 瓶颈类型: cpu/memory/network/disk */
  type: string
  /** 严重程度: critical/high/medium/low */
  severity: string
  /** 详细描述 */
  description: string
  /** 当前指标值 */
  metric_value: number
  /** 阈值 */
  threshold: number
}

/**
 * 业务趋势类型
 */
export interface BusinessTrend {
  /** 指标名称 */
  metric_name: string
  /** 趋势方向: up/down/stable */
  direction: string
  /** 变化百分比 */
  change_percent: number
  /** 起始值 */
  start_value: number
  /** 结束值 */
  end_value: number
  /** 时间序列数据点列表 */
  data_points: Record<string, any>[]
}

/**
 * 优化建议类型
 */
export interface OptimizationSuggestion {
  /** 优先级: high/medium/low */
  priority: string
  /** 建议分类，如 资源优化/配置调整/架构改进 */
  category: string
  /** 建议内容 */
  suggestion: string
  /** 预期影响 */
  impact: string
}

/**
 * 指标统计数据类型
 */
export interface MetricStats {
  /** 最小值 */
  min?: number
  /** 最大值 */
  max?: number
  /** 平均值 */
  avg?: number
  /** 趋势变化率 */
  trend?: number
}

/**
 * 集群性能指标类型
 */
export interface ClusterHealth {
  /** 整体状态: healthy/warning/critical */
  status: string
  /** 性能评分（0-100） */
  score?: number
  /** 性能状态详细分析 */
  analysis?: string
  /** CPU 平均使用率百分比 */
  cpu_usage_pct?: number
  /** 内存平均使用率百分比 */
  memory_usage_pct?: number
  /** 磁盘平均使用率百分比 */
  disk_usage_pct?: number
  /** 带宽平均使用率百分比 */
  network_bandwidth_pct?: number
  /** 系统平均负载 */
  load_average?: number
  /** 网络流量（Mbps） */
  network_traffic?: number
  /** 错误率百分比 */
  error_rate?: number
  /** 瓶颈数量 */
  bottleneck_count?: number
  /** 高优先级瓶颈数量 */
  high_count?: number
  /** 中优先级瓶颈数量 */
  medium_count?: number
  /** 低优先级瓶颈数量 */
  low_count?: number
  /** CPU 详细统计 */
  cpu_stats?: MetricStats
  /** 内存详细统计 */
  memory_stats?: MetricStats
  /** 磁盘详细统计 */
  disk_stats?: MetricStats
  /** 网络详细统计 */
  network_stats?: Record<string, any>
  /** 负载详细统计 */
  load_stats?: MetricStats
  /** 分析时间段 */
  analysis_period?: string
  /** 时间戳 */
  timestamp?: string
}

/**
 * 完整性能分析报告类型
 */
export interface PerformanceReport {
  /** 报告唯一标识符 */
  id: string
  /** 创建时间 */
  created_at: string
  /** 分析时长（小时） */
  analysis_period_hours: number
  /** 命名空间，为 null 表示全集群 */
  namespace: string | null
  /** 报告状态: completed/basic/failed/analyzing */
  status: string
  /** 报告摘要 */
  summary: string
  /** 集群性能指标 */
  cluster_health: ClusterHealth | null
  /** 性能瓶颈列表 */
  bottlenecks: PerformanceBottleneck[]
  /** 业务趋势列表 */
  trends: BusinessTrend[]
  /** 优化建议列表 */
  suggestions: OptimizationSuggestion[]
  /** 原始摘要（未格式化） */
  raw_summary: string | null
}

/**
 * 分析任务触发请求类型
 */
export interface AnalysisTriggerRequest {
  /** 分析时长（小时），默认 24 小时 */
  analysis_period_hours?: number
  /** 命名空间，不填则分析全集群 */
  namespace?: string | null
}

/**
 * 分析任务状态类型
 */
export interface AnalysisTaskStatus {
  /** 任务唯一标识符 */
  task_id: string
  /** 任务状态: queued/running/completed/failed */
  status: string
  /** 任务进度（0-100） */
  progress: number
  /** 创建时间 */
  created_at: string
  /** 更新时间 */
  updated_at: string
  /** 生成的报告 ID（任务完成时） */
  report_id: string | null
  /** 错误信息（任务失败时） */
  error_message: string | null
}

/**
 * 报告列表响应类型
 */
export interface ReportListResponse {
  /** 总报告数 */
  total: number
  /** 当前页码 */
  page: number
  /** 每页数量 */
  page_size: number
  /** 报告列表 */
  reports: PerformanceReport[]
}

/**
 * 定时任务配置类型
 */
export interface ScheduleConfig {
  /** 是否启用定时任务 */
  enabled: boolean
  /** 执行时间（小时，0-23），默认凌晨 2 点 */
  hour: number
  /** 执行时间（分钟，0-59），默认 0 分 */
  minute: number
  /** 分析时长（小时），默认 24 小时 */
  analysis_period_hours: number
}

/**
 * 定时任务状态类型
 */
export interface ScheduleStatus {
  /** 定时任务是否正在运行 */
  is_running: boolean
  /** 下次执行时间 */
  next_run: string | null
  /** 定时配置 */
  config: ScheduleConfig
  /** 配置文件路径 */
  config_file?: string
}

/**
 * 业务分析 API 调用模块
 */
export const analysisApi = {
  /**
   * 触发分析任务
   * @param data 分析触发请求参数
   * @returns 包含 task_id 的响应
   */
  triggerAnalysis: (data: AnalysisTriggerRequest) => {
    return request.post<any, ApiResponse<{ task_id: string }>>('/v1/analysis/trigger', data)
  },

  /**
   * 查询任务状态
   * @param taskId 任务唯一标识符
   * @returns 任务状态信息
   */
  getTaskStatus: (taskId: string) => {
    return request.get<any, ApiResponse<AnalysisTaskStatus>>(`/v1/analysis/task/${taskId}`)
  },

  /**
   * 查询最近一次任务状态
   * @returns 最近任务状态信息
   */
  getLatestTask: () => {
    return request.get<any, ApiResponse<AnalysisTaskStatus>>('/v1/analysis/task/latest')
  },

  /**
   * 分页查询报告列表
   * @param page 页码，默认 1
   * @param pageSize 每页数量，默认 10
   * @returns 分页报告列表
   */
  getReports: (page?: number, pageSize?: number) => {
    const params: Record<string, number> = {}
    if (page !== undefined) {
      params.page = page
    }
    if (pageSize !== undefined) {
      params.page_size = pageSize
    }
    return request.get<any, ApiResponse<ReportListResponse>>('/v1/analysis/reports', { params })
  },

  /**
   * 获取报告详情
   * @param reportId 报告唯一标识符
   * @returns 报告详情
   */
  getReport: (reportId: string) => {
    return request.get<any, ApiResponse<PerformanceReport>>(`/v1/analysis/reports/${reportId}`)
  },

  /**
   * 删除报告
   * @param reportId 报告唯一标识符
   * @returns 删除结果
   */
  deleteReport: (reportId: string) => {
    return request.delete<any, ApiResponse<{ report_id: string }>>(`/v1/analysis/reports/${reportId}`)
  },

  /**
   * 获取定时任务状态
   * @returns 定时任务状态
   */
  getScheduleStatus: () => {
    return request.get<any, ApiResponse<ScheduleStatus>>('/v1/analysis/schedule/status')
  },

  /**
   * 启动定时任务
   * @returns 启动结果
   */
  startSchedule: () => {
    return request.post<any, ApiResponse<{ success: boolean }>>('/v1/analysis/schedule/start')
  },

  /**
   * 停止定时任务
   * @returns 停止结果
   */
  stopSchedule: () => {
    return request.post<any, ApiResponse<{ success: boolean }>>('/v1/analysis/schedule/stop')
  },

  /**
   * 更新定时配置
   * @param config 定时配置
   * @returns 更新后的配置
   */
  updateScheduleConfig: (config: ScheduleConfig) => {
    return request.put<any, ApiResponse<ScheduleConfig>>('/v1/analysis/schedule/config', config)
  }
}

File: agent/backend/api/routes/analysis.py
"""
业务分析相关路由
提供触发分析任务、查询任务状态、管理报告、配置定时任务等 API 接口
"""

import logging

from fastapi import APIRouter, HTTPException, status, Query

from api.schemas import (
    ApiResponse,
    AnalysisTriggerRequest,
    AnalysisTaskStatus,
    PerformanceReport,
    ReportListResponse,
    ScheduleConfig,
)
from agents.analysis_task_manager import get_analysis_task_manager
from agents.report_store import get_report_store
from agents.schedule_analyzer import get_schedule_analyzer


logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/analysis", tags=["业务分析"])


@router.post(
    "/trigger",
    response_model=ApiResponse[dict],
    status_code=status.HTTP_202_ACCEPTED,
    summary="手动触发分析任务"
)
async def trigger_analysis(request: AnalysisTriggerRequest) -> ApiResponse[dict]:
    """
    手动触发业务分析任务（异步执行）
    
    Args:
        request: 分析触发请求，包含分析时长和命名空间
        
    Returns:
        ApiResponse[dict]: 202 Accepted，包含 task_id
    """
    try:
        logger.info(
            f"触发分析任务: analysis_period_hours={request.analysis_period_hours}, "
            f"namespace={request.namespace or '全集群'}"
        )
        
        # 获取分析任务管理器并触发分析
        task_manager = get_analysis_task_manager()
        task_id = task_manager.trigger_analysis(
            period_hours=request.analysis_period_hours,
            namespace=request.namespace
        )
        
        logger.info(f"分析任务已触发: task_id={task_id}")
        
        return ApiResponse[dict](
            code=202,
            message="已接受，分析任务已启动",
            data={"task_id": task_id},
        )
        
    except Exception as e:
        logger.error(f"触发分析任务失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"触发分析任务失败: {str(e)}",
        )


@router.get(
    "/task/latest",
    response_model=ApiResponse[AnalysisTaskStatus],
    status_code=status.HTTP_200_OK,
    summary="查询最近一次任务状态"
)
async def get_latest_task() -> ApiResponse[AnalysisTaskStatus]:
    """
    查询最近一次触发的分析任务状态
    
    Returns:
        ApiResponse[AnalysisTaskStatus]: 最近任务状态信息，data 为 null 表示没有任务记录
    """
    try:
        logger.info("查询最近一次任务状态")
        
        # 获取分析任务管理器并查询最近任务
        task_manager = get_analysis_task_manager()
        task_info = task_manager.get_latest_task()
        
        # 如果没有任务记录，返回 200 但 data 为 null
        if task_info is None:
            logger.info("没有找到任何任务记录，返回空响应")
            return ApiResponse[AnalysisTaskStatus](
                code=200,
                message="没有找到任何任务记录",
                data=None,
            )
        
        # 构建任务状态响应
        task_status = AnalysisTaskStatus(
            task_id=task_info["task_id"],
            status=task_info["status"],
            progress=task_info["progress"],
            created_at=task_info["created_at"],
            updated_at=task_info["updated_at"],
            report_id=task_info.get("report_id"),
            error_message=task_info.get("error_message"),
        )
        
        logger.info(
            f"最近任务查询成功: task_id={task_status.task_id}, "
            f"status={task_status.status}"
        )
        
        return ApiResponse[AnalysisTaskStatus](
            code=200,
            message="success",
            data=task_status,
        )
        
    except Exception as e:
        logger.error(f"查询最近任务状态失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"查询最近任务状态失败: {str(e)}",
        )


@router.get(
    "/task/{task_id}",
    response_model=ApiResponse[AnalysisTaskStatus],
    status_code=status.HTTP_200_OK,
    summary="查询指定任务状态"
)
async def get_task_status(task_id: str) -> ApiResponse[AnalysisTaskStatus]:
    """
    根据任务 ID 查询分析任务状态
    
    Args:
        task_id: 任务唯一标识符
        
    Returns:
        ApiResponse[AnalysisTaskStatus]: 任务状态信息
        
    Raises:
        HTTPException: 任务不存在时返回 404
    """
    try:
        logger.info(f"查询任务状态: task_id={task_id}")
        
        # 获取分析任务管理器并查询任务状态
        task_manager = get_analysis_task_manager()
        task_info = task_manager.get_task_status(task_id)
        
        if task_info is None:
            logger.warning(f"任务不存在: task_id={task_id}")
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"任务不存在: {task_id}",
            )
        
        # 构建任务状态响应
        task_status = AnalysisTaskStatus(
            task_id=task_info["task_id"],
            status=task_info["status"],
            progress=task_info["progress"],
            created_at=task_info["created_at"],
            updated_at=task_info["updated_at"],
            report_id=task_info.get("report_id"),
            error_message=task_info.get("error_message"),
        )
        
        logger.info(
            f"任务状态查询成功: task_id={task_id}, status={task_status.status}, "
            f"progress={task_status.progress}%"
        )
        
        return ApiResponse[AnalysisTaskStatus](
            code=200,
            message="success",
            data=task_status,
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"查询任务状态失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"查询任务状态失败: {str(e)}",
        )


@router.get(
    "/reports",
    response_model=ApiResponse[ReportListResponse],
    status_code=status.HTTP_200_OK,
    summary="分页查询报告列表"
)
async def list_reports(
    page: int = Query(default=1, ge=1, description="页码，从 1 开始"),
    page_size: int = Query(default=10, ge=1, le=100, description="每页数量，最大 100"),
) -> ApiResponse[ReportListResponse]:
    """
    分页查询性能分析报告列表（按时间倒序）
    
    Args:
        page: 页码，默认 1
        page_size: 每页数量，默认 10
        
    Returns:
        ApiResponse[ReportListResponse]: 分页报告列表
    """
    try:
        logger.info(f"查询报告列表: page={page}, page_size={page_size}")
        
        # 获取报告存储器并分页查询
        report_store = get_report_store()
        result = report_store.list(page=page, page_size=page_size)
        
        # 构建报告列表响应
        reports = []
        for report_data in result["reports"]:
            report = PerformanceReport(**report_data)
            reports.append(report)
        
        report_list = ReportListResponse(
            total=result["total"],
            page=result["page"],
            page_size=result["page_size"],
            reports=reports,
        )
        
        logger.info(
            f"报告列表查询成功: total={report_list.total}, "
            f"returned={len(report_list.reports)}"
        )
        
        return ApiResponse[ReportListResponse](
            code=200,
            message="success",
            data=report_list,
        )
        
    except Exception as e:
        logger.error(f"查询报告列表失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"查询报告列表失败: {str(e)}",
        )


@router.get(
    "/reports/{report_id}",
    response_model=ApiResponse[PerformanceReport],
    status_code=status.HTTP_200_OK,
    summary="查询报告详情"
)
async def get_report(report_id: str) -> ApiResponse[PerformanceReport]:
    """
    根据报告 ID 查询性能分析报告详情
    
    Args:
        report_id: 报告唯一标识符
        
    Returns:
        ApiResponse[PerformanceReport]: 报告详情
        
    Raises:
        HTTPException: 报告不存在时返回 404
    """
    try:
        logger.info(f"查询报告详情: report_id={report_id}")
        
        # 获取报告存储器并查询报告
        report_store = get_report_store()
        report_data = report_store.get(report_id)
        
        if report_data is None:
            logger.warning(f"报告不存在: report_id={report_id}")
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"报告不存在: {report_id}",
            )
        
        # 构建报告响应
        report = PerformanceReport(**report_data)
        
        logger.info(f"报告查询成功: report_id={report_id}")
        
        return ApiResponse[PerformanceReport](
            code=200,
            message="success",
            data=report,
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"查询报告详情失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"查询报告详情失败: {str(e)}",
        )


@router.delete(
    "/reports/{report_id}",
    response_model=ApiResponse[dict],
    status_code=status.HTTP_200_OK,
    summary="删除报告"
)
async def delete_report(report_id: str) -> ApiResponse[dict]:
    """
    根据报告 ID 删除性能分析报告
    
    Args:
        report_id: 报告唯一标识符
        
    Returns:
        ApiResponse[dict]: 成功消息
        
    Raises:
        HTTPException: 报告不存在时返回 404
    """
    try:
        logger.info(f"删除报告: report_id={report_id}")
        
        # 获取报告存储器并删除报告
        report_store = get_report_store()
        success = report_store.delete(report_id)
        
        if not success:
            logger.warning(f"报告不存在，无法删除: report_id={report_id}")
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"报告不存在: {report_id}",
            )
        
        logger.info(f"报告删除成功: report_id={report_id}")
        
        return ApiResponse[dict](
            code=200,
            message="报告删除成功",
            data={"report_id": report_id},
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除报告失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"删除报告失败: {str(e)}",
        )


@router.get(
    "/schedule/status",
    response_model=ApiResponse[dict],
    status_code=status.HTTP_200_OK,
    summary="查询定时任务状态"
)
async def get_schedule_status() -> ApiResponse[dict]:
    """
    查询定时分析任务的配置和运行状态
    
    Returns:
        ApiResponse[dict]: 包含 ScheduleConfig 和运行状态
    """
    try:
        logger.info("查询定时任务状态")
        
        # 获取定时分析调度器并查询状态
        schedule_analyzer = get_schedule_analyzer()
        status_info = schedule_analyzer.get_status()
        
        # 构建调度配置
        schedule_config = ScheduleConfig(
            enabled=status_info["enabled"],
            hour=status_info["schedule_time"]["hour"],
            minute=status_info["schedule_time"]["minute"],
            analysis_period_hours=status_info["analysis_period_hours"],
        )
        
        # 构建完整的状态响应
        response_data = {
            "is_running": status_info["is_running"],
            "next_run": status_info.get("next_run"),
            "config": schedule_config.model_dump(),
            "config_file": status_info.get("config_file"),
        }
        
        logger.info(
            f"定时任务状态查询成功: is_running={response_data['is_running']}, "
            f"enabled={schedule_config.enabled}"
        )
        
        return ApiResponse[dict](
            code=200,
            message="success",
            data=response_data,
        )
        
    except Exception as e:
        logger.error(f"查询定时任务状态失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"查询定时任务状态失败: {str(e)}",
        )


@router.post(
    "/schedule/start",
    response_model=ApiResponse[dict],
    status_code=status.HTTP_200_OK,
    summary="启动定时任务"
)
async def start_schedule() -> ApiResponse[dict]:
    """
    启动定时分析任务调度器
    
    Returns:
        ApiResponse[dict]: 操作结果
    """
    try:
        logger.info("启动定时任务")
        
        # 获取定时分析调度器并启动
        schedule_analyzer = get_schedule_analyzer()
        success = schedule_analyzer.start()
        
        if success:
            logger.info("定时任务启动成功")
            return ApiResponse[dict](
                code=200,
                message="定时任务启动成功",
                data={"success": True},
            )
        else:
            logger.warning("定时任务已经在运行")
            return ApiResponse[dict](
                code=200,
                message="定时任务已经在运行",
                data={"success": False},
            )
        
    except Exception as e:
        logger.error(f"启动定时任务失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"启动定时任务失败: {str(e)}",
        )


@router.post(
    "/schedule/stop",
    response_model=ApiResponse[dict],
    status_code=status.HTTP_200_OK,
    summary="停止定时任务"
)
async def stop_schedule() -> ApiResponse[dict]:
    """
    停止定时分析任务调度器
    
    Returns:
        ApiResponse[dict]: 操作结果
    """
    try:
        logger.info("停止定时任务")
        
        # 获取定时分析调度器并停止
        schedule_analyzer = get_schedule_analyzer()
        success = schedule_analyzer.stop()
        
        if success:
            logger.info("定时任务停止成功")
            return ApiResponse[dict](
                code=200,
                message="定时任务停止成功",
                data={"success": True},
            )
        else:
            logger.warning("定时任务未在运行")
            return ApiResponse[dict](
                code=200,
                message="定时任务未在运行",
                data={"success": False},
            )
        
    except Exception as e:
        logger.error(f"停止定时任务失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"停止定时任务失败: {str(e)}",
        )


@router.put(
    "/schedule/config",
    response_model=ApiResponse[ScheduleConfig],
    status_code=status.HTTP_200_OK,
    summary="更新定时配置"
)
async def update_schedule_config(request: ScheduleConfig) -> ApiResponse[ScheduleConfig]:
    """
    更新定时分析任务的配置
    
    Args:
        request: 定时配置请求
        
    Returns:
        ApiResponse[ScheduleConfig]: 更新后的配置
        
    Raises:
        HTTPException: 更新失败时返回 500
    """
    try:
        logger.info(
            f"更新定时配置: hour={request.hour}, minute={request.minute}, "
            f"period_hours={request.analysis_period_hours}, enabled={request.enabled}"
        )
        
        # 获取定时分析调度器并更新配置
        schedule_analyzer = get_schedule_analyzer()
        success = schedule_analyzer.update_config(
            hour=request.hour,
            minute=request.minute,
            period_hours=request.analysis_period_hours,
            enabled=request.enabled,
        )
        
        if not success:
            logger.error("更新定时配置失败")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="更新定时配置失败，请检查参数是否有效",
            )
        
        # 获取更新后的配置
        status_info = schedule_analyzer.get_status()
        updated_config = ScheduleConfig(
            enabled=status_info["enabled"],
            hour=status_info["schedule_time"]["hour"],
            minute=status_info["schedule_time"]["minute"],
            analysis_period_hours=status_info["analysis_period_hours"],
        )
        
        logger.info("定时配置更新成功")
        
        return ApiResponse[ScheduleConfig](
            code=200,
            message="定时配置更新成功",
            data=updated_config,
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新定时配置失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"更新定时配置失败: {str(e)}",
        )
